File: packages/oxlint-plugin-react-doctor/src/plugin/rules/a11y/shadcn-form-item-requires-label.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { getTrailingJsxNameSegment } from "../../utils/get-trailing-jsx-name-segment.js";
import { hasJsxPropIgnoreCase } from "../../utils/has-jsx-prop-ignore-case.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import {
  SHADCN_UI_MODULE_SOURCE_PATTERN,
  resolveShadcnUiComponentName,
} from "../../utils/resolve-shadcn-ui-component-name.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { visitStaticJsxChildren } from "../../utils/visit-static-jsx-children.js";
import { walkAst } from "../../utils/walk-ast.js";

const FORM_MODULE_PATTERN = /(?:^|\/)ui\/(?:.*\/)?form$|^\.\.?\/(?:.*\/)?form$/;
const LABEL_ATTRIBUTES = ["aria-label", "aria-labelledby"] as const;

// react-hook-form's render prop hands the control its wiring as
// `{ field }` — `{...field}` spreads name/value/onChange/onBlur/ref and
// never an aria label, so the canonical shadcn form field stays checkable.
// Any other spread may deliver aria-label and makes the claim unprovable.
const RHF_FIELD_SPREAD_NAME_PATTERN = /^field(?:$|[A-Z])/;

const hasUnknownSpreadAttribute = (attributes: ReadonlyArray<EsTreeNode>): boolean =>
  attributes.some(
    (attribute) =>
      isNodeOfType(attribute, "JSXSpreadAttribute") &&
      !(
        isNodeOfType(attribute.argument, "Identifier") &&
        RHF_FIELD_SPREAD_NAME_PATTERN.test(attribute.argument.name)
      ),
  );

// shadcn's FormLabel wires htmlFor to the FormControl's generated id — a
// FormItem whose control has no FormLabel (and no aria naming) renders an
// unlabeled field. Intrinsic `label` and a name-alike local wrapper also
// satisfy the claim.
const isLabelElementName = (elementName: EsTreeNode, context: RuleContext): boolean => {
  if (resolveShadcnUiComponentName(elementName, FORM_MODULE_PATTERN, context) === "FormLabel") {
    return true;
  }
  const trailingSegment = getTrailingJsxNameSegment(elementName);
  return (
    trailingSegment === "label" || trailingSegment === "FormLabel" || trailingSegment === "Label"
  );
};

interface FormItemScan {
  hasControl: boolean;
  hasLabel: boolean;
  sawUnprovableContent: boolean;
}

const isInsideFormControl = (
  node: EsTreeNode,
  formItemElement: EsTreeNodeOfType<"JSXElement">,
  context: RuleContext,
): boolean => {
  let ancestor = node.parent;
  while (ancestor && ancestor !== formItemElement) {
    if (
      isNodeOfType(ancestor, "JSXElement") &&
      resolveShadcnUiComponentName(ancestor.openingElement.name, FORM_MODULE_PATTERN, context) ===
        "FormControl"
    ) {
      return true;
    }
    ancestor = ancestor.parent;
  }
  return false;
};

const scanFormItem = (
  formItemElement: EsTreeNodeOfType<"JSXElement">,
  context: RuleContext,
): FormItemScan => {
  const scan: FormItemScan = { hasControl: false, hasLabel: false, sawUnprovableContent: false };
  visitStaticJsxChildren(formItemElement.children, {
    onElement: (element) => {
      const openingElement = element.openingElement;
      const elementName = openingElement.name;
      if (isLabelElementName(elementName, context)) {
        scan.hasLabel = true;
        return false;
      }
      const resolvedPartName = resolveShadcnUiComponentName(
        elementName,
        FORM_MODULE_PATTERN,
        context,
      );
      if (
        (resolvedPartName === "FormControl" ||
          isInsideFormControl(openingElement, formItemElement, context)) &&
        LABEL_ATTRIBUTES.some((attribute) =>
          hasJsxPropIgnoreCase(openingElement.attributes, attribute),
        )
      ) {
        // The control (or a wrapper) names itself directly.
        scan.hasLabel = true;
        return false;
      }
      if (hasUnknownSpreadAttribute(openingElement.attributes)) {
        // A spread can deliver aria-label to the control at runtime.
        scan.sawUnprovableContent = true;
      }
      if (resolvedPartName === "FormControl") {
        scan.hasControl = true;
        return true;
      }
      const trailingSegment = getTrailingJsxNameSegment(elementName);
      const isCustomComponent =
        trailingSegment !== null &&
        /^[A-Z]/.test(trailingSegment) &&
        trailingSegment !== "Fragment";
      if (
        isCustomComponent &&
        resolveShadcnUiComponentName(elementName, SHADCN_UI_MODULE_SOURCE_PATTERN, context) === null
      ) {
        // An unresolved custom component may render the label itself.
        scan.sawUnprovableContent = true;
      }
      return true;
    },
    onOpaqueExpression: (expression) => {
      scan.sawUnprovableContent = true;
      // Labels written inside map callbacks or other opaque expressions are
      // still statically visible.
      walkAst(expression, (node) => {
        if (isNodeOfType(node, "JSXOpeningElement") && isLabelElementName(node.name, context)) {
          scan.hasLabel = true;
        }
      });
    },
  });
  return scan;
};

export const shadcnFormItemRequiresLabel = defineRule({
  id: "shadcn-form-item-requires-label",
  title: "FormItem control without a label",
  severity: "warn",
  requires: ["shadcn"],
  recommendation:
    "Give every FormItem that wraps a FormControl a FormLabel (sr-only when the design shows no visible label) or name the control directly with aria-label.",
  create: (context: RuleContext) => ({
    JSXOpeningElement(node: EsTreeNodeOfType<"JSXOpeningElement">) {
      if (resolveShadcnUiComponentName(node.name, FORM_MODULE_PATTERN, context) !== "FormItem") {
        return;
      }
      const formItemElement = node.parent;
      if (
        !formItemElement ||
        !isNodeOfType(formItemElement, "JSXElement") ||
        formItemElement.children.length === 0
      ) {
        return;
      }
      const scan = scanFormItem(formItemElement, context);
      if (!scan.hasControl || scan.hasLabel || scan.sawUnprovableContent) return;
      context.report({
        node: node.name,
        message:
          "This FormItem wraps a FormControl but renders no FormLabel, so the field has no accessible name. Add a FormLabel (visually hidden if the design shows no label) or an aria-label on the control.",
      });
    },
  }),
});

File: native/oxlint/rules/shadcn-form-item-requires-label.rs
use std::cell::Cell;

use oxc_ast::{ast::JSXAttributeItem, AstKind};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_span::GetSpan;

use crate::{context::LintContext, rule::Rule, utils::has_jsx_prop_ignore_case, AstNode};

const MESSAGE: &str = "This FormItem wraps a FormControl but renders no FormLabel, so the field has no accessible name. Add a FormLabel (visually hidden if the design shows no label) or an aria-label on the control.";
const LABEL_ATTRIBUTES: [&str; 2] = ["aria-label", "aria-labelledby"];

#[derive(Debug, Default, Clone)]
pub struct ShadcnFormItemRequiresLabel;

declare_oxc_lint!(
    /// Require accessible labels on shadcn form items.
    ShadcnFormItemRequiresLabel,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Require accessible labels on shadcn form items.",
);

impl Rule for ShadcnFormItemRequiresLabel {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        if !has_capability_or_unspecified(ctx, "shadcn") {
            return;
        }
        let AstKind::JSXOpeningElement(opening_element) = node.kind() else {
            return;
        };
        if resolve_shadcn_component_name(&opening_element.name, "form", ctx).as_deref()
            != Some("FormItem")
        {
            return;
        }
        let parent = ctx.nodes().parent_node(node.id());
        let AstKind::JSXElement(form_item) = parent.kind() else {
            return;
        };
        if form_item.children.is_empty() {
            return;
        }
        let has_control = Cell::new(false);
        let has_label = Cell::new(false);
        let saw_unprovable_content = Cell::new(false);
        visit_static_jsx_children(
            &form_item.children,
            &mut |element| {
                let child_opening_element = &element.opening_element;
                if is_shadcn_form_label_name(&child_opening_element.name, ctx) {
                    has_label.set(true);
                    return false;
                }
                let resolved_part =
                    resolve_shadcn_component_name(&child_opening_element.name, "form", ctx);
                if (resolved_part.as_deref() == Some("FormControl")
                    || is_inside_shadcn_form_control(child_opening_element, form_item, ctx))
                    && has_form_label_attribute(child_opening_element)
                {
                    has_label.set(true);
                    return false;
                }
                if has_unknown_form_spread(child_opening_element) {
                    saw_unprovable_content.set(true);
                }
                if resolved_part.as_deref() == Some("FormControl") {
                    has_control.set(true);
                    return true;
                }
                if resolved_part.is_none()
                    && jsx_element_name_trailing_segment(&child_opening_element.name).is_some_and(
                        |name| {
                            name != "Fragment"
                                && name.as_bytes().first().is_some_and(u8::is_ascii_uppercase)
                                && resolve_general_shadcn_ui_component_name(
                                    &child_opening_element.name,
                                    ctx,
                                )
                                .is_none()
                        },
                    )
                {
                    saw_unprovable_content.set(true);
                }
                true
            },
            &mut || saw_unprovable_content.set(true),
            &mut || {},
        );
        if !has_control.get() || has_label.get() || saw_unprovable_content.get() {
            return;
        }
        ctx.diagnostic(OxcDiagnostic::warn(MESSAGE).with_label(opening_element.name.span()));
    }
}

fn is_shadcn_form_label_name<'a>(
    element_name: &oxc_ast::ast::JSXElementName<'a>,
    ctx: &LintContext<'a>,
) -> bool {
    resolve_shadcn_component_name(element_name, "form", ctx).as_deref() == Some("FormLabel")
        || matches!(
            jsx_element_name_trailing_segment(element_name),
            Some("label" | "FormLabel" | "Label")
        )
}

fn has_form_label_attribute(opening_element: &oxc_ast::ast::JSXOpeningElement<'_>) -> bool {
    LABEL_ATTRIBUTES
        .iter()
        .any(|attribute_name| has_jsx_prop_ignore_case(opening_element, attribute_name).is_some())
}

fn has_unknown_form_spread(opening_element: &oxc_ast::ast::JSXOpeningElement<'_>) -> bool {
    opening_element.attributes.iter().any(|attribute| {
        let JSXAttributeItem::SpreadAttribute(spread_attribute) = attribute else {
            return false;
        };
        let oxc_ast::ast::Expression::Identifier(identifier) =
            spread_attribute.argument.get_inner_expression()
        else {
            return true;
        };
        !is_react_hook_form_field_name(identifier.name.as_str())
    })
}

fn is_react_hook_form_field_name(identifier_name: &str) -> bool {
    identifier_name == "field"
        || identifier_name
            .strip_prefix("field")
            .and_then(|suffix| suffix.as_bytes().first())
            .is_some_and(u8::is_ascii_uppercase)
}

fn is_inside_shadcn_form_control<'a>(
    opening_element: &oxc_ast::ast::JSXOpeningElement<'a>,
    form_item: &oxc_ast::ast::JSXElement<'a>,
    ctx: &LintContext<'a>,
) -> bool {
    for ancestor in ctx.nodes().ancestors(opening_element.node_id.get()) {
        if ancestor.id() == form_item.node_id.get() {
            return false;
        }
        if matches!(
            ancestor.kind(),
            AstKind::JSXElement(element)
                if resolve_shadcn_component_name(&element.opening_element.name, "form", ctx)
                    .as_deref()
                    == Some("FormControl")
        ) {
            return true;
        }
    }
    false
}
